### PR TITLE
Issue #861 Fix slow services

### PR DESCRIPTION
**Pull request by bensinjin**
This PR adds an expiry to topic services and ensures we only fetch services from the server once every 24 hours for a given topic.

### CODE REVIEW

File: src/components/services/service_list_component.tsx [hunk 1]
```diff
@@ -42,83 +39,77 @@ export interface ServicesUpdater {
...
+
+const refreshServices = (props: Props): void => {
+    if (servicesFetchRequired(props.topicServicesOrError)) {
+        props.dispatchServicesRequest();
+    }
+};
+
+const servicesFetchRequired = (topicServicesOrError: SelectorTopicServices): boolean => (
+    !(topicServicesOrError.type === constants.TOPIC_SERVICES_VALID && topicServicesOrError.isStaleData === false)
+);
```
Comment (rasmus-storjohann-PG) [inline src/components/services/service_list_component.tsx]: Nice short functions. Can be made more readable by the rule `!(A && B) === !A || !B`. Generally avoid `=== false` and `=== true`

File: src/components/services/service_list_component.tsx [hunk 2]
```diff
@@ -42,83 +39,77 @@ export interface ServicesUpdater {
...
         return topicServicesOrError.errorMessageType;
     }
     return Errors.Exception;
 };
 
-const isLoadingServices = (topicServicesOrError: SelectorTopicServices): boolean => (
-    topicServicesOrError.type === constants.TOPIC_SERVICES_LOADING
+const getServices = (topicServicesOrError: SelectorTopicServices): ReadonlyArray<HumanServiceData> => (
+    topicServicesOrError.type === constants.TOPIC_SERVICES_VALID && topicServicesOrError.services
 );
```
Comment (rasmus-storjohann-PG) [inline src/components/services/service_list_component.tsx]: I think renaming this function to `getServicesIfValid()` will make us think of the invalid case at call sites and squash bugs.

File: src/selectors/services/is_error_selector_topic_services.ts [hunk 3]
```diff
@@ -0,0 +1,7 @@
+import { SelectorTopicServices, ErrorSelectorTopicServices } from './types';
+import * as constants from '../../application/constants';
+
+export const isErrorSelectorTopicServices = (topicServicesOrError: SelectorTopicServices):
+    topicServicesOrError is ErrorSelectorTopicServices => (
+        (<ErrorSelectorTopicServices>topicServicesOrError).type === constants.TOPIC_SERVICES_ERROR
+    );
```
Comment (rasmus-storjohann-PG) [inline src/selectors/services/is_error_selector_topic_services.ts]: Have a look at the code with this function just called `isError()`. Does it cause ambiguity at the call sites, or would you say that the context makes it clear what the function is doing? If making this rename, do the same to the isValid case.

File: src/selectors/services/to_valid_selector_topic_services.ts [hunk 4]
```diff
@@ -6,4 +6,5 @@ export const toValidSelectorTopicServices = (topicServices: ValidServicesForTopi
     ValidSelectorTopicServices => ({
         services: topicServices.serviceIds.map((serviceId: ServiceId) => services[serviceId]),
         type: constants.TOPIC_SERVICES_VALID,
+        isStaleData: topicServices.expiresAt < Date.now(),
```
Comment (rasmus-storjohann-PG) [inline src/selectors/services/to_valid_selector_topic_services.ts]: Avoid introducing multiple terms for the same thing, so `isExpired`.

File: src/validation/services/__tests__/to_human_service_data.test.ts [hunk 5]
```diff
@@ -65,64 +65,36 @@ describe('Adapting server service object to client service object', () => {
...
-                    country: serverServiceObject.location.addresses[0].address.country,
-                }],
-                website: serverServiceObject.service.organization_url,
-                email: serverServiceObject.service.organization_email,
-                organizationName: serverServiceObject.service.organization_name,
-                bookmarked: true,
-            };
-            expect(clientServiceObject[0]).toEqual(expectedClientServiceObject);
-        });
+    it('builds expected structure with "bookmarked" set to true when bookmark exists', () => {
```
Comment (rasmus-storjohann-PG) [inline src/validation/services/__tests__/to_human_service_data.test.ts]: There are two tests here, and they're testing for three things: build the expected structure, sets bookmarked to false, and sets bookmarked to true. As they stand, both test checks for the expected structure, which is why there is duplication. Clearly there should be three tests, only one of them should contain a bunch of JSON.

File: src/validation/services/index.ts [hunk 6]
```diff
@@ -53,47 +53,42 @@ export interface ValidatedServiceAtLocationJSON {
...
+    ReadonlyArray<types.HumanServiceData> => (
+        R.map(serviceFromData(bookmarkedServiceIds), data)
+    );
 
-export const serviceFromValidatedJSON = (data: ValidatedServiceAtLocationJSON): types.HumanServiceData => {
-    const phoneNumbers = R.map((phoneNumber: ValidatedPhoneNumberJSON): types.PhoneNumber => ({
-        type: phoneNumber.phone_number_type,
-        phone_number: phoneNumber.phone_number,
-    }), data.location.phone_numbers);
+const serviceFromData = R.curry((bookmarkedServiceIds: ReadonlyArray<Id>, data: ValidatedServiceAtLocationJSON):
```
Comment (rasmus-storjohann-PG) [inline src/validation/services/index.ts]: This function would benefit from being broken up into several helper functions.

File: src/stores/__tests__/services.test.ts [hunk 7]
```diff
@@ -139,6 +139,38 @@ describe('services reducer', () => {
...
+                    topicId,
+                    services: [serviceBuilder.build()],
+                },
+            };
+            const storeAfterAction = reducer(storeBeforeAction, theAction);
+            const servicesByTopic = storeAfterAction.servicesByTopic[topicId];
+
+            if (isValidServicesForTopic(servicesByTopic)) {
+                const twentyFourHoursAfterExpiry = expiresAt + (24 * 60 * 60 * 1000);
+                expect(servicesByTopic.expiresAt).toBeGreaterThanOrEqual(twentyFourHoursAfterExpiry);
```
[the remark this answers is not in the document]
Comment (bensinjin) [reply src/stores/__tests__/services.test.ts]: Turns out we can actually test for the exact value here which I've updated this test to do.

File: src/components/services/service_list_component.tsx (hunk 1 above)
Comment (bensinjin) [reply src/components/services/service_list_component.tsx]: Sounds good - I've cleaned this up to add clarity and removed the `=== false`.

File: src/components/services/service_list_component.tsx (hunk 2 above)
Comment (bensinjin) [reply src/components/services/service_list_component.tsx]: Fixed

File: src/stores/__tests__/services.test.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (bensinjin) [reply src/stores/__tests__/services.test.ts]: I stand corrected, this works locally but we get some latency when CI runs it - I've added an upper limit.

File: src/selectors/services/is_error_selector_topic_services.ts (hunk 3 above)
Comment (bensinjin) [reply src/selectors/services/is_error_selector_topic_services.ts]: This was renamed to better reflect that we're referring to the selector type as is `isValidSelectorTopicServices`. There exists another version of these that accepts the store's  types so I felt it was preferable to be specific.

File: src/selectors/services/to_valid_selector_topic_services.ts (hunk 4 above)
Comment (bensinjin) [reply src/selectors/services/to_valid_selector_topic_services.ts]: fixed

File: src/validation/services/__tests__/to_human_service_data.test.ts (hunk 5 above)
Comment (bensinjin) [reply src/validation/services/__tests__/to_human_service_data.test.ts]: fixed

File: src/validation/services/index.ts (hunk 6 above)
Comment (bensinjin) [reply src/validation/services/index.ts]: fixed

**Comment by bensinjin** [comment]
@rasmus-storjohann-PG I think this is ready

**Comment by rasmus-storjohann-PG** [comment]
Looking at it now, it dropped off my radar a bit

**Pull request merged**